ark/phenotyping/cluster.py

```diff
@@ -0,0 +1,151 @@
+        winning_coords (tuple):
+            A coordinate array which indicates the winning neuron's position
+        sigma (float):
+            Determines the spread of the Gaussian neighborhood function
+        learning_rate (float):
+            Determines how sensitive the weight updates will be to new data
+        x_mesh (numpy.ndarray):
+            The x coordinate matrix of the weights vectors
+        y_mesh (numpy.ndarray):
+            The y coordinate matrix of the weights vectors
+        t (int):
+            The current iteration index
+        num_iters (int):
+            The maximum number of iterations
+
+    Returns:
+        numpy.ndarray:
+            The updated weights matrix
+    """
+
+    # update learning rate with asymptotic decay
+    decay_lr = learning_rate / (1 + t / (num_iters / 2))
+
+    # update sigma with asymptotic decay
+    decay_sig = sigma / (1 + t / (num_iters / 2))
+
+    # return a Gaussian centered around the winning coordinates
+    d = 2 * np.pi * decay_sig * decay_sig
+    ax = np.exp(-np.power(x_mesh - x_mesh.T[winning_coords], 2) / d)
+    ay = np.exp(-np.power(y_mesh - y_mesh.T[winning_coords], 2) / d)
+    g = (ax * ay).T * decay_lr
+
+    # update the weights based on the Gaussian neighborhood
+    weights += np.einsum('ij, ijk->ijk', g, sample - weights)
+
+    return weights
+
+
+def train_flowsom(pixel_mat, x_neurons, y_neurons, num_iters,
+                  sigma=1.0, learning_rate=0.5, random_seed=0):
+    """Trains the SOM by iterating through the each data point and updating the params
+
+    Args:
+        pixel_mat (pandas.DataFrame):
+            A matrix with pixel-level channel information for non-zero pixels in img_xr
+        x_neurons (int):
+            The number of x neurons to use
+        y_neurons (int):
+            The number of y neurons to use
+        num_iters (int):
+            The maximum number of iterations for training
+        sigma (float):
+            Determines the spread of the Gaussian neighborhood function
+        learning_rate (float):
+            Determines how sensitive the weight updates will be to new data
+        random_seed (int):
+            The seed to set for random weight initialization
+
+    Returns:
+        numpy.ndarray:
+            The weights matrix after training
+    """
+
+    # define the random generator
+    rand_gen = np.random.RandomState(random_seed)
+
+    # initialize the weights and normalize
+    weights = rand_gen.rand(x_neurons, y_neurons, pixel_mat.shape[1]) * 2 - 1
+    weights /= np.linalg.norm(weights, axis=-1, keepdims=True)
+
+    # define the activation map
+    activation_map = np.zeros((x_neurons, y_neurons))
+
+    # define meshgrid coords for the weights matrix, convert to float (yikes, memory...)
+    x_mesh, y_mesh = np.meshgrid(np.arange(x_neurons), np.arange(y_neurons))
+    x_mesh = x_mesh.astype(float)
+    y_mesh = y_mesh.astype(float)
+
```

Comment:
if the floats here are a big deal, we can just use `scipy.stats.mulitvariate_distribution` to generate `g` in the update function.  This reduces from 2*100^2 floats to either that many uint8's or no meshgrid at all (read through the `multivariate_distribution_docs` to figure that part out)

Reply:
Now that I think about it, there's no real reason why converting to float type is bad in terms of memory. We're only ever going to be setting x and y to 10, so 100 floats should absolutely be able to be accommodated. However, if there's no difference in correctness between int and a float, definitely better to leave it as int and not have the extra conversion step.